torchprime/torch_xla_models/trainer/sft_trainer.py

```diff
@@ -0,0 +1,155 @@
+"""Trainer for supervised fine-tuning (SFT) tasks."""
+
+from __future__ import annotations
+
+import logging
+import multiprocessing as mp
+import tempfile
+import time
+from pathlib import Path
+
+import torch
+import torch.distributed as dist
+import torch.distributed.checkpoint as dist_cp
+import torch_xla.core.xla_model as xm
+import torch_xla.experimental.distributed_checkpoint as xc
+import torch_xla.runtime as xr
+from omegaconf import DictConfig
+from torch import nn
+from torch.distributed.checkpoint import FileSystemReader, FileSystemWriter
+
+from torchprime.torch_xla_models.model.base_causal_lm import (
+  save_sharded_safetensors_by_layer,
+)
+
+from .base_trainer import Trainer
+
+logger = logging.getLogger(__name__)
+
+
+class SFTTrainer(Trainer):
+  """Trainer with pretrained weight loading and saving support."""
+
+  def __init__(
+    self,
+    model: nn.Module,
+    config: DictConfig,
+    train_dataset,
+  ) -> None:
+    """Initialize trainer and optionally load pretrained weights.
+
+    Args:
+      model: Model instance to train.
+      config: Hydra configuration object.
+      train_dataset: Dataset used for training.
+    """
+
+    self.pretrained_model = getattr(config.model, "pretrained_model", None)
+
+    if self.pretrained_model:
+      if xr.process_index() == 0:
+        logger.info("Loading model weights from %s", self.pretrained_model)
+      model.from_pretrained(self.pretrained_model)
+      xm.mark_step()
+    else:
+      logger.info(
+        "No pretrained model specified; training from scratch. \n\nIs this what you intended?\n"
+      )
+
+    super().__init__(model, config, train_dataset)
+
+  def train_loop(self) -> None:
+    """Run the base training loop and export the model.
+
+    Args:
+      metrics_logger: Instance used to record metrics during training.
+    """
+    super().train_loop()
+
+    t0 = time.perf_counter()
+    logger.info("[SAVING] Starting distributed checkpoint …")
+    self._maybe_save_model()
+    dt = time.perf_counter() - t0
+    logger.info("[SAVING] Finished in %.2f s", dt)
+
+  def _maybe_save_model(self) -> None:
```

Comment:
it doesn't look like this method should belong to sft_trainer, could you extract all _maybe logic and move it to base model next to save_sharded_safetensors_by_layer or to some util module